xrmocap/core/evaluation/end2end_evaluation.py

```diff
@@ -22,111 +28,315 @@ class End2EndEvaluation(BaseEvaluation):
+                        frame_idx, :n_valid_person] = keypoints3d_pred_valid
+
+            pred_keypoints3d_raw = Keypoints(
+                dtype='numpy',
+                kps=pred_kps3d,
+                mask=pred_kps3d[..., -1] > 0,
+                convention=self.pred_kps3d_convention,
                 logger=self.logger)
-            pred_keypoints3d = convert_keypoints(
-                pred_keypoints3d,
-                dst=self.eval_kps3d_convention,
-                approximate=True)
-            scene_gt_kps3d = gt_kps3d[start_idx:end_idx]
-            gt_keypoints3d = Keypoints(
-                kps=scene_gt_kps3d,
-                mask=scene_gt_kps3d[..., -1] > 0,
-                convention=self.dataset.gt_kps3d_convention,
+
+            # prepare gt Keypoints3d
+            gt_n_person = np.max(
+                np.array([
+                    scene_keypoints.get_person_number()
+                    for scene_keypoints in self.dataset.gt3d
+                ]))
+            gt_n_kps = self.dataset.gt3d[0].get_keypoints_number()
+            gt_kps3d = np.full((self.n_frame, gt_n_person, gt_n_kps, 4),
+                               np.nan)
+            gt_mask = np.full((self.n_frame, gt_n_person, gt_n_kps), np.nan)
+
+            start_frame = 0
+            for gt_keypoints3d_scene in self.dataset.gt3d:
+
+                gt_kps3d_scene = gt_keypoints3d_scene.get_keypoints().numpy(
+                )  # [n_frame, n_person, n_kps, 4]
+                gt_mask_scene = gt_keypoints3d_scene.get_mask().numpy(
+                )  # [n_frame, n_person, n_kps]
+
+                n_frame, n_person, _, _ = gt_kps3d_scene.shape
+                end_frame = min(start_frame + n_frame, self.n_frame)
+
+                gt_kps3d[start_frame:end_frame, :n_person, ...] = \
+                    gt_kps3d_scene[:end_frame-start_frame, ...]
+                gt_mask[start_frame:end_frame, :n_person, ...] = \
+                    gt_mask_scene[:end_frame-start_frame, ...]
+                start_frame = end_frame
+
+            if self.dataset_name == 'panoptic' and \
+                    self.trans_matrix is not None:
+                for frame_idx in range(self.n_frame):
+                    for person_idx in range(gt_n_person):
+                        gt_kps3d[frame_idx, person_idx, :, :3] = \
+                            gt_kps3d[frame_idx, person_idx, :, :3].dot(
+                                self.trans_matrix)
+
+            gt_keypoints3d_raw = Keypoints(
+                dtype='numpy',
+                kps=gt_kps3d,
+                mask=gt_mask,
+                convention=self.gt_kps3d_convention,
                 logger=self.logger)
-            gt_keypoints3d = convert_keypoints(
-                gt_keypoints3d,
-                dst=self.eval_kps3d_convention,
-                approximate=True)
-            eval_results = self.metric_manager(
+
+            # convert pred and gt to the same convention before passing
```

Comment:
Shall we move the process about convention and intersection mask into `metric_manager`? It can be reused by other evaluation.

Reply:
Added function `align_convention_mask` in `xrmocap.utils.eval_utils` in [2342328](https://github.com/openxrlab/xrmocap/pull/79/commits/2342328e10b30abba07cf90b328d1a78375812cb) for better reuse in other evaluation classes. Since the scope of the metric manager is to deal with various metrics and their calculation, it assumes the input keypoints are aligned already. It is better to do the keypoints alignment outside, before passing them to the metric manager.